### PR TITLE
Implement tree-like fold function for ShareArray

### PR DESCRIPTION
**Pull request by Drake-Eidukas**
Share Equality and Comparison both require a tree-like fold for maximum
efficiency. This implements it on the ShareArray class to be as DRY as
possible. This allows operations such as summing and taking the product sum of a sharearray to proceed as parallel as possible.

### CODE REVIEW

**Comment by Drake-Eidukas** [comment]
This also fixes a bug where equality could break if the security parameter was odd, as the zip operation truncates the last element of x if x had an odd length.

File: honeybadgermpc/progs/mixins/dataflow.py [hunk 1]
```diff
@@ -228,6 +228,50 @@ def __div__(self, other):
...
+
+        returns:
+            A Share or ShareFuture representing the iterated binary operation of op on
+            the shares of this array
+        """
+        shares = self._shares
+        assert len(shares) > 0
+
+        while len(shares) > 1:
+            results = [op(a, b) for a, b in zip(shares[::2], shares[1::2])]
```
Comment (rahulgovind) [inline honeybadgermpc/progs/mixins/dataflow.py]: Unfortunately, I had observed that share reconstructions do not work very well when done in parallel. The reconstructions need to be executed in the same order across all nodes, and I believe this condition can be violated when multiple reconstructions are carried out in parallel. 

Additionally, there is a pretty high overhead with calling the operations on individual shares. Instead, for each level, you can apply the operation on arrays instead. I have versions of this in https://github.com/rahulgovind/HoneyBadgerMPC/blob/fixed-point-order/honeybadgermpc/fixedpoint.py. binary_addition has a prefix sum, get_carry_bits has a reduction.

**Review by rahulgovind — CHANGES_REQUESTED** [review]

File: honeybadgermpc/progs/mixins/dataflow.py (hunk 1 above)
Comment (Drake-Eidukas) [reply honeybadgermpc/progs/mixins/dataflow.py]: 1. Not sure how to address this point, as I don't think this changes the way we do share reconstruction for the arrays... but maybe I'm misunderstanding your point.
2. That's a good idea-- I'll update this to do that.

Comment (Drake-Eidukas) [reply honeybadgermpc/progs/mixins/dataflow.py]: Yep, about 10 seconds faster for an array with 1000 elements using number 2! good catch :D

Comment (rahulgovind) [reply honeybadgermpc/progs/mixins/dataflow.py]: I'm not too sure how point 1 can be addressed either. Just that the await statement is required for now. The changes look great! How much time does it take to do the 1000 element array operation right now?

**Review by rahulgovind — APPROVED** [review]

File: honeybadgermpc/progs/mixins/dataflow.py (hunk 1 above)
Comment (Drake-Eidukas) [reply honeybadgermpc/progs/mixins/dataflow.py]: ~So, the total time to run all the tests in the file with preprocessing 10000 elements took about 38 seconds with this new version, the old version took 53 seconds for the whole file, which includes the same operation 3 times on 80 elements for the equality test. not a very scientific approach to testing the time... I'd keep the test in but I was having difficulty with `test_preprocessing` to get the test case of 1000 elements to work when I ran the tests over all of the files (the tests worked perfectly when only running on this one file... but there's issues where test_preprocessing will cache calls to generate more elements but then not actually reset the elements in the preprocessing to contain enough elements. Perhaps I can add it back in once that gets refactored).~
With the exact amount of preprocessing required, it takes ~1.3 seconds on my machine

**Pull request merged**